ngraph/core/reference/include/ngraph/runtime/reference/adaptive_avg_pool.hpp

```diff
@@ -0,0 +1,169 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <cfenv>
+#include <cmath>
+#include <numeric>
+#include <stdexcept>
+#include <vector>
+
+#include "ngraph/axis_vector.hpp"
+#include "ngraph/shape.hpp"
+
+namespace ngraph
+{
+    namespace runtime
+    {
+        namespace reference
+        {
+            namespace adaptive_pool
+            {
+                inline size_t window_start(size_t idx, size_t arg_shape, size_t out_shape)
+                {
+                    return floor(static_cast<double>(idx * arg_shape) / out_shape);
+                }
+                inline size_t window_end(size_t idx, size_t arg_shape, size_t out_shape)
+                {
+                    return ceil(static_cast<double>((idx + 1) * arg_shape) / out_shape);
+                }
+                template <typename T>
+                T avg(const T sum, size_t n)
+                {
+                    if (n == 0)
+                    {
+                        NGRAPH_CHECK(n != 0, "AdaptiveAvgPool elements == 0, must be non-zero");
+                    }
+
+                    if (std::is_same<T, int8_t>::value || std::is_same<T, uint8_t>::value)
+                    {
+                        return static_cast<T>(std::nearbyint(static_cast<float>(sum) / n));
+                    }
+                    else
+                    {
+                        return sum / n;
+                    }
+                }
+
+                template <typename T>
+                void adaptive_avg_pool_1d(const T* arg, T* out, size_t h_in, size_t h_out)
+                {
+                    for (size_t i = 0; i < h_out; i++)
+                    {
+                        size_t h_start = window_start(i, h_in, h_out);
+                        size_t h_end = window_end(i, h_in, h_out);
+                        out[i] =
+                            avg(std::accumulate(arg + h_start, arg + h_end, T{0}), h_end - h_start);
+                    }
+                }
+                template <typename T>
+                void adaptive_avg_pool_2d(
+                    const T* arg, T* out, size_t h_in, size_t h_out, size_t w_in, size_t w_out)
+                {
+                    for (size_t i = 0; i < h_out; i++)
+                    {
+                        size_t h_start = window_start(i, h_in, h_out);
+                        size_t h_end = window_end(i, h_in, h_out);
+                        for (size_t j = 0; j < w_out; j++)
+                        {
+                            size_t w_start = window_start(j, w_in, w_out);
+                            size_t w_end = window_end(j, w_in, w_out);
+                            T result = 0;
+                            for (size_t n = h_start; n < h_end; n++)
+                                result = std::accumulate(
+                                    arg + n * w_in + w_start, arg + n * w_in + w_end, result);
```

Comment:
Consider to use always perenties in for loop even if it's only one line:
```suggestion
                            for (size_t n = h_start; n < h_end; n++)
                            {
                                const auto b =  arg + n * w_in + w_start;
                                const auto e = arg + n * w_in + w_end;
                                result = std::accumulate(b, e, result);
                             }
```